## **1. 데이터셋 불러오기**

In [1]:
import seaborn as sns

titanic = sns.load_dataset('titanic')

In [2]:
titanic

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

## **2. feature 분석**
### **해당 데이터의 feature 파악

### **2-1. head 함수를 이용해 데이터 프레임의 첫 5행을 출력하여 어떤 feature들이 있는지 확인해주세요.**

In [3]:
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### **2-2. describe 함수를 통해서 기본적인 통계를 확인해주세요.**

In [4]:
titanic.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

### **2-3. describe 함수를 통해 확인할 수 있는 count, std, min, 25%, 50%, 70%, max 가 각각 무슨 뜻인지 주석 혹은 markdown 블록으로 간단히 설명해주세요.**

#### **count**
##### 각 열에 포함된 데이터의 개수. 결측치가 없는 경우 총 데이터 수를 나타냅니다.

#### **mean**
##### 평균값. 데이터의 전반적인 경향을 보여줍니다.

#### **std**
##### 표준편차. 데이터가 평균으로부터 얼마나 퍼져 있는지를 나타냅니다. 값이 클수록 데이터의 변동성이 크다는 의미입니다.

#### **min**
##### 최소값. 해당 열에서 가장 작은 값을 나타냅니다.

#### **25%**
##### 1사분위수. 데이터의 하위 25%에 해당하는 값으로, 데이터의 첫 번째 사분위수를 나타냅니다.

#### **50%**
##### 중앙값(중위수). 데이터의 중간값으로, 데이터의 절반이 이 값보다 작고 나머지 절반이 이 값보다 큽니다.

#### **75%**
##### 3사분위수. 데이터의 상위 25%에 해당하는 값으로, 데이터의 세 번째 사분위수를 나타냅니다.

#### **max**
##### 최대값. 해당 열에서 가장 큰 값을 나타냅니다.

### **2-4. isnull() 함수와 sum()  함수를 이용해 각 열의 결측치 갯수를 확인해주세요.**

In [5]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## **3. feature engineering**
### **모델의 성능 향상을 위한 단계**

### **3-1. 결측치 처리**
#### **Age(나이)의 결측치는 중앙값으로, Embarked(승선 항구)의 결측치는 최빈값으로 대체해주세요. 모두 대체한 후에, 대체 결과를 isnull() 함수와 sum()  함수를 이용해서 확인해주세요.**

In [6]:
# 나이 결측치를 중앙값으로 대체
titanic['age'] = titanic['age'].fillna(titanic['age'].median())

# 탑승지 결측치를 최빈값으로 대체
titanic['embarked'] = titanic['embarked'].fillna(titanic['embarked'].mode()[0])

In [7]:
titanic['age'].isnull().sum()

0

In [8]:
titanic['embarked'].isnull().sum()

0

### **3-2. 수치형으로 인코딩**
#### **Sex(성별)를 남자는 0, 여자는 1로 변환해주세요. alive(생존여부)를 yes는 1, no는 0으로 변환해주세요. Embarked(승선 항구)는 ‘C’는 0으로, Q는 1으로, ‘S’는 2로 변환해주세요. 모두 변환한 후에, 변환 결과를 head 함수를 이용해 확인해주세요.**

In [9]:
titanic['sex'] = titanic['sex'].map({'male': 0, 'female': 1})
titanic['alive'] = titanic['alive'].map({'no': 1, 'yes': 0})
titanic['embarked'] = titanic['embarked'].map({'C': 0, 'Q': 1, 'S': 2,})

In [10]:
titanic['sex'].head()

0    0
1    1
2    1
3    1
4    0
Name: sex, dtype: int64

In [11]:
titanic['alive'].head()

0    1
1    0
2    0
3    0
4    1
Name: alive, dtype: int64

In [12]:
titanic['embarked'].head()

0    2
1    0
2    2
3    2
4    2
Name: embarked, dtype: int64

### **3-2.  새로운 feature 생성**
#### **SibSip(타이타닉호에 동승한 자매 및 배우자의 수), Parch(타이타닉호에 동승한 부모 및 자식의 수)를 통해서 family_size(가족크기)를 생성해주세요. 새로운 feature를 head 함수를 이용해 확인해주세요.**

In [13]:
titanic['family_size'] = titanic['sibsp'] + titanic['parch'] + 1

In [14]:
print(titanic['family_size'].head())

0    2
1    2
2    1
3    2
4    1
Name: family_size, dtype: int64


In [15]:
titanic

survived  pclass  sex   age  sibsp  parch     fare  embarked   class  \
0           0       3    0  22.0      1      0   7.2500         2   Third   
1           1       1    1  38.0      1      0  71.2833         0   First   
2           1       3    1  26.0      0      0   7.9250         2   Third   
3           1       1    1  35.0      1      0  53.1000         2   First   
4           0       3    0  35.0      0      0   8.0500         2   Third   
..        ...     ...  ...   ...    ...    ...      ...       ...     ...   
886         0       2    0  27.0      0      0  13.0000         2  Second   
887         1       1    1  19.0      0      0  30.0000         2   First   
888         0       3    1  28.0      1      2  23.4500         2   Third   
889         1       1    0  26.0      0      0  30.0000         0   First   
890         0       3    0  32.0      0      0   7.7500         1   Third   

       who  adult_male deck  embark_town  alive  alone  family_size  
0      man        True  NaN  Southampton      1  False            2  
1    woman       False    C    Cherbourg      0  False            2  
2    woman       False  NaN  Southampton      0   True            1  
3    woman       False    C  Southampton      0  False            2  
4      man        True  NaN  Southampton      1   True            1  
..     ...         ...  ...          ...    ...    ...          ...  
886    man        True  NaN  Southampton      1   True            1  
887  woman       False    B  Southampton      0   True            1  
888  woman       False  NaN  Southampton      1  False            4  
889    man        True    C    Cherbourg      0   True            1  
890    man        True  NaN   Queenstown      1   True            1  

[891 rows x 16 columns]

## **4. 모델 학습시키기 (Logistic Regression, Random Forest, XGBoost)** 

### **4-1. 모델 학습 준비**
#### **이제 모델을 학습시키기 위한 데이터를 준비하겠습니다. 학습에 필요한 feature은 'survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', ‘family_size’ 입니다. feature과 target을 분리해주세요.  그 다음 데이터 스케일링을 진행해주세요.**
##### **feature들의 값을 일정한 범위로 맞추는 과정**

###### **데이터프레임 분석에 필요한 열들만 선택 후 feature와 Target 분리:**

In [16]:
titanic = titanic[['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'family_size']]
X = titanic.drop('survived', axis=1) # feature
y = titanic['survived'] # target

In [17]:
titanic

survived  pclass  sex   age  sibsp  parch     fare  embarked  family_size
0           0       3    0  22.0      1      0   7.2500         2            2
1           1       1    1  38.0      1      0  71.2833         0            2
2           1       3    1  26.0      0      0   7.9250         2            1
3           1       1    1  35.0      1      0  53.1000         2            2
4           0       3    0  35.0      0      0   8.0500         2            1
..        ...     ...  ...   ...    ...    ...      ...       ...          ...
886         0       2    0  27.0      0      0  13.0000         2            1
887         1       1    1  19.0      0      0  30.0000         2            1
888         0       3    1  28.0      1      2  23.4500         2            4
889         1       1    0  26.0      0      0  30.0000         0            1
890         0       3    0  32.0      0      0   7.7500         1            1

[891 rows x 9 columns]

### **4-2. Logistic Regression : 로지스틱 회귀 기법을 사용.*
#### ***문제에서는 분리후 스케일링을 진행해달라고 했지만 데이터 분할이후 스케일링을 진행하는 것이 대체로 옳다고 구글링돼서 데이터 분할 이후 스케일링을 진행하였다**

In [18]:
from sklearn.model_selection import train_test_split

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

##### **random_state=42 부가설명 : 난수 발생 시 시드값을 제어합니다** 
###### **쉽게 설명하면 , 데이터를 분할할 때마다 다른 방식으로 나누어 진다면 결과가 매번 달라질 수 있으므로, 결과의 일관성이 유지되도록 도와주는 매개변수 입니다.**

In [19]:
X

pclass  sex   age  sibsp  parch     fare  embarked  family_size
0         3    0  22.0      1      0   7.2500         2            2
1         1    1  38.0      1      0  71.2833         0            2
2         3    1  26.0      0      0   7.9250         2            1
3         1    1  35.0      1      0  53.1000         2            2
4         3    0  35.0      0      0   8.0500         2            1
..      ...  ...   ...    ...    ...      ...       ...          ...
886       2    0  27.0      0      0  13.0000         2            1
887       1    1  19.0      0      0  30.0000         2            1
888       3    1  28.0      1      2  23.4500         2            4
889       1    0  26.0      0      0  30.0000         0            1
890       3    0  32.0      0      0   7.7500         1            1

[891 rows x 8 columns]

In [20]:
y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 891, dtype: int64

In [21]:
from sklearn.preprocessing import StandardScaler

# 데이터 스케일링
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [22]:
titanic

survived  pclass  sex   age  sibsp  parch     fare  embarked  family_size
0           0       3    0  22.0      1      0   7.2500         2            2
1           1       1    1  38.0      1      0  71.2833         0            2
2           1       3    1  26.0      0      0   7.9250         2            1
3           1       1    1  35.0      1      0  53.1000         2            2
4           0       3    0  35.0      0      0   8.0500         2            1
..        ...     ...  ...   ...    ...    ...      ...       ...          ...
886         0       2    0  27.0      0      0  13.0000         2            1
887         1       1    1  19.0      0      0  30.0000         2            1
888         0       3    1  28.0      1      2  23.4500         2            4
889         1       1    0  26.0      0      0  30.0000         0            1
890         0       3    0  32.0      0      0   7.7500         1            1

[891 rows x 9 columns]

In [23]:
import numpy as np
from sklearn.preprocessing import StandardScaler

# 원본 데이터 저장
X_train_original = X_train.copy()  # 스케일링 전 원본 데이터 저장

# 데이터 스케일링
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 스케일링 전의 평균과 표준편차
print("스케일링 전:")
print("X_train 평균:", np.mean(X_train_original, axis=0))
print("X_train 표준편차:", np.std(X_train_original, axis=0))

# 스케일링 후의 평균과 표준편차
print("\n스케일링 후:")
print("X_train 평균:", np.mean(X_train, axis=0))
print("X_train 표준편차:", np.std(X_train, axis=0))

# 스케일링된 데이터 샘플 출력
print("\n스케일링된 X_train 데이터 샘플:")
print(X_train[:5])  # 첫 5행만 출력



스케일링 전:
X_train 평균: [-4.11655728e-17 -6.51788236e-17 -1.54370898e-17 -2.40678264e-16
  2.73189711e-16  3.86706896e-17 -2.68199944e-16 -1.39606257e-16]
X_train 표준편차: [1. 1. 1. 1. 1. 1. 1. 1.]

스케일링 후:
X_train 평균: [-1.49692992e-17  2.49488320e-17  1.24744160e-17  3.49283648e-17
 -4.24130144e-17  1.74641824e-17 -1.49692992e-17 -3.74232480e-17]
X_train 표준편차: [1. 1. 1. 1. 1. 1. 1. 1.]

스케일링된 X_train 데이터 샘플:
[[-1.61413602 -0.7243102   1.25364106 -0.47072241 -0.47934164 -0.07868358
   0.5635246  -0.55466613]
 [-0.40055118 -0.7243102  -0.47728355 -0.47072241 -0.47934164 -0.37714494
   0.5635246  -0.55466613]
 [ 0.81303367 -0.7243102   0.21508629 -0.47072241 -0.47934164 -0.47486697
   0.5635246  -0.55466613]
 [ 0.81303367 -0.7243102  -0.24649361  0.37992316 -0.47934164 -0.47623026
   0.5635246   0.04009635]
 [ 0.81303367  1.38062393 -1.78509326  2.93185988  2.04874166 -0.02524937
   0.5635246   3.01390875]]


In [24]:
from sklearn.linear_model import LogisticRegression

# 모델 생성 및 학습
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [25]:
# 예측
y_pred = model.predict(X_test)

**예측의 의미**
- X_test: 모델이 학습하지 않은 새로운 데이터입니다. 일반적으로 이 데이터는 모델의 성능을 평가하기 위해 사용됩니다.
- y_pred: X_test의 각 샘플에 대해 모델이 예측한 결과를 저장하는 배열입니다. 이 배열에는 각 샘플에 대한 예측 클래스(예: 0 또는 1)가 담기게 됩니다.

In [26]:
from sklearn.metrics import accuracy_score, classification_report

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")

Accuracy: 0.8044692737430168
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.86      0.84       105
           1       0.78      0.73      0.76        74

    accuracy                           0.80       179
   macro avg       0.80      0.79      0.80       179
weighted avg       0.80      0.80      0.80       179



# 모델 평가 결과

## 1. Accuracy
- **Accuracy**: 0.804 (약 80.4%)
  - 전체 데이터에서 모델이 올바르게 예측한 비율입니다. 이 경우, 전체 179개의 샘플 중 약 80.4%를 정확하게 분류했습니다.

## 2. Classification Report
보고서에서 제공하는 주요 지표는 다음과 같습니다:

### Precision
- 0의 Precision: 0.82
- 1의 Precision: 0.78
  - Precision은 모델이 양성으로 예측한 것 중 실제 양성의 비율입니다.
  - 0의 Precision이 0.82는 모델이 0으로 예측한 것 중 82%가 실제로도 0이라는 의미입니다.
  - 1의 Precision이 0.78은 모델이 1로 예측한 것 중 78%가 실제로 1이라는 의미입니다.

### Recall
- 0의 Recall: 0.86
- 1의 Recall: 0.73
  - Recall은 실제 양성 중에서 모델이 올바르게 양성으로 예측한 비율입니다.
  - 0의 Recall이 0.86이라는 것은 실제 0인 샘플 중 86%를 올바르게 예측했다는 것을 의미합니다.
  - 1의 Recall이 0.73이라는 것은 실제 1인 샘플 중 73%를 올바르게 예측했다는 의미입니다.

### F1-Score
- 0의 F1-Score: 0.84
- 1의 F1-Score: 0.76
  - F1-Score는 Precision과 Recall의 조화 평균으로, 두 지표의 균형을 고려합니다.
  - 0의 F1-Score가 0.84로 양호하며, 1의 F1-Score는 0.76으로 다소 낮습니다.

### Support
- 0의 Support: 105 (실제 0인 샘플 수)
- 1의 Support: 74 (실제 1인 샘플 수)
  - Support는 각 클래스에 대한 실제 샘플 수를 나타냅니다.

## 3. Macro Average 및 Weighted Average
- **Macro Avg**: 
  - 각 클래스의 Precision, Recall, F1-Score의 평균을 단순히 계산한 값입니다. 모든 클래스에 동일한 가중치를 부여합니다.
  
- **Weighted Avg**: 
  - 각 클래스의 Precision, Recall, F1-Score를 Support 수에 따라 가중 평균한 값입니다. 클래스 비율이 다른 경우 더 현실적인 평가를 제공합니다.

## 종합 해석
- **생존자 수**: 모델의 결과를 바탕으로, 총 179명의 샘플 중 74명이 생존자(1 클래스)로 확인되었습니다.
- **모델 성능**: 모델은 전체적으로 80.4%의 정확도를 보여주며, 이는 양호한 수준입니다.
- **클래스 간 성능 차이**: 0 클래스에 대한 성능(Precision, Recall, F1-Score)이 1 클래스보다 전반적으로 좋습니다. 특히 1 클래스의 Recall이 0.73으로 낮아, 실제 1인 경우를 잘 예측하지 못하는 경향이 있습니다. 
- **개선 방향**: 1 클래스의 성능을 향상시키기 위해 추가적인 데이터 수집, 데이터 전처리, 모델 조정 등을 고려해볼 수 있습니다.


--------------------

In [27]:
from sklearn.ensemble import RandomForestClassifier

# 랜덤포레스트 모델 학습
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

# 예측 및 평가
y_pred_rf = rf.predict(X_test)
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print(classification_report(y_test, y_pred_rf))

Random Forest Accuracy: 0.8212290502793296
              precision    recall  f1-score   support

           0       0.85      0.85      0.85       105
           1       0.78      0.78      0.78        74

    accuracy                           0.82       179
   macro avg       0.82      0.82      0.82       179
weighted avg       0.82      0.82      0.82       179



In [28]:
from sklearn.tree import DecisionTreeClassifier  # 의사결정나무 모델 임포트

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 스케일링
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 모델 생성 및 학습
model = DecisionTreeClassifier(random_state=42)
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")

Accuracy: 0.770949720670391
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.76      0.80       105
           1       0.70      0.78      0.74        74

    accuracy                           0.77       179
   macro avg       0.77      0.77      0.77       179
weighted avg       0.78      0.77      0.77       179



In [29]:
import xgboost as xgb
from sklearn.metrics import mean_squared_error

# 데이터 분할
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 데이터 스케일링
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# XGBoost 모델 생성
xgb_model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# 모델 학습
xgb_model.fit(X_train_scaled, y_train)

# 예측
y_pred_xgb = xgb_model.predict(X_test_scaled)

# 평가
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
print(f'XGBoost 모델의 MSE: {mse_xgb}')

XGBoost 모델의 MSE: 0.12981004899201257


In [30]:
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

# 모델 생성 및 학습 #나이브 베이즈
model = GaussianNB()
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_test, y_pred)}")

Accuracy: 0.7821229050279329
Classification Report:
              precision    recall  f1-score   support

           0       0.83      0.79      0.81       105
           1       0.72      0.77      0.75        74

    accuracy                           0.78       179
   macro avg       0.78      0.78      0.78       179
weighted avg       0.79      0.78      0.78       179

Confusion Matrix:
[[83 22]
 [17 57]]


In [31]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# 모델 생성 및 학습 SVM
model = SVC(kernel='linear')
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_test, y_pred)}")

Accuracy: 0.7821229050279329
Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.84      0.82       105
           1       0.75      0.70      0.73        74

    accuracy                           0.78       179
   macro avg       0.78      0.77      0.77       179
weighted avg       0.78      0.78      0.78       179

Confusion Matrix:
[[88 17]
 [22 52]]


In [32]:
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score

# XGBoost 분류 모델 학습
model = XGBClassifier()
model.fit(X_train, y_train)

# 예측 수행
y_pred = model.predict(X_test)

# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print(f"XGBoost Model Accuracy: {accuracy}")


XGBoost Model Accuracy: 0.7932960893854749


In [33]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# 모델 생성 및 학습 KNN
model = KNeighborsClassifier(n_neighbors=5)
model.fit(X_train, y_train)

# 예측
y_pred = model.predict(X_test)

# 평가
print(f"Accuracy: {accuracy_score(y_test, y_pred)}")
print(f"Classification Report:\n{classification_report(y_test, y_pred)}")
print(f"Confusion Matrix:\n{confusion_matrix(y_test, y_pred)}")


Accuracy: 0.7094972067039106
Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.84      0.77       105
           1       0.70      0.53      0.60        74

    accuracy                           0.71       179
   macro avg       0.71      0.68      0.69       179
weighted avg       0.71      0.71      0.70       179

Confusion Matrix:
[[88 17]
 [35 39]]


In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import BaggingClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.impute import SimpleImputer

In [35]:
# 배깅 모델 생성
bagging_model = BaggingClassifier(estimator=DecisionTreeClassifier(), n_estimators=100, random_state=42)

# 모델 학습
bagging_model.fit(X_train, y_train)

# 예측
y_pred_bagging = bagging_model.predict(X_test)

# 평가
accuracy_bagging = accuracy_score(y_test, y_pred_bagging)
print(f'배깅 모델의 정확도: {accuracy_bagging}')
print(classification_report(y_test, y_pred_bagging))

# ☑️ 부스팅 모델 구현 및 평가
# 부스팅 모델 생성
boosting_model = GradientBoostingClassifier(n_estimators=100, random_state=42)

# 모델 학습
boosting_model.fit(X_train, y_train)

# 예측
y_pred_boosting = boosting_model.predict(X_test)

# 평가
accuracy_boosting = accuracy_score(y_test, y_pred_boosting)
print(f'부스팅 모델의 정확도: {accuracy_boosting}')
print(classification_report(y_test, y_pred_boosting))

배깅 모델의 정확도: 0.8100558659217877
              precision    recall  f1-score   support

           0       0.83      0.86      0.84       105
           1       0.79      0.74      0.76        74

    accuracy                           0.81       179
   macro avg       0.81      0.80      0.80       179
weighted avg       0.81      0.81      0.81       179

부스팅 모델의 정확도: 0.8044692737430168
              precision    recall  f1-score   support

           0       0.81      0.88      0.84       105
           1       0.80      0.70      0.75        74

    accuracy                           0.80       179
   macro avg       0.80      0.79      0.79       179
weighted avg       0.80      0.80      0.80       179



In [36]:
# GBM 모델 생성
gbm_model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# 모델 학습
gbm_model.fit(X_train, y_train)

# 예측
y_pred_gbm = gbm_model.predict(X_test)

# 평가
accuracy_gbm = accuracy_score(y_test, y_pred_gbm)
print(f'GBM 모델의 정확도: {accuracy_gbm}')
print(classification_report(y_test, y_pred_gbm))

GBM 모델의 정확도: 0.8044692737430168
              precision    recall  f1-score   support

           0       0.81      0.88      0.84       105
           1       0.80      0.70      0.75        74

    accuracy                           0.80       179
   macro avg       0.80      0.79      0.79       179
weighted avg       0.80      0.80      0.80       179



## 결론

모델 성능 평가 결과는 다음과 같습니다 (정확도 높은 순서):

1. **랜덤 포레스트(Random Forest)**: 정확도 82.12% (0.8212290502793296)
2. **배깅(Bagging)**: 정확도 81.01% (0.8100558659217877)
3. **부스팅(Boosting)**: 정확도 80.45% (0.8044692737430168)
4. **GBM 모델**: 정확도 80.45% (0.8044692737430168)
5. **XGBoost 모델**: 정확도 79.33% (0.7932960893854749)
6. **나이브 베이즈(Naive Bayes)**: 정확도 78.21% (0.7821229050279329)
7. **서포트 벡터 머신(SVM)**: 정확도 78.21% (0.7821229050279329)
8. **의사 결정 나무(Decision Tree)**: 정확도 77.09% (0.770949720670391)
9. **K-최근접 이웃(KNN)**: 정확도 70.95% (0.7094972067039106)

위의 결과를 종합적으로 분석해보면, **랜덤 포레스트 모델이 82.12%의 정확도로 가장 뛰어난 성능을 보였습니다.** 이는 랜덤 포레스트가 여러 개의 의사 결정 나무를 앙상블하여 데이터의 복잡한 패턴을 더 잘 학습했음을 나타냅니다.

다른 모델들과 비교했을 때, 랜덤 포레스트의 성능은 확실히 우수하며, 이는 다양한 상황에서 효과적인 선택이 될 수 있음을 시사합니다.